Refcoder CFrame Special Cases & Fix

ArvidSilverlock opened the pull request:
Implemented [CFrame special cases](https://dom.rojo.space/binary.html#cframe) to reduce pack sizes, done via adding an extra byte to the packed value, each representing a CFrame's rotation (zero being a non-special case). Thus allowing CFrames with certain rotations to be stored the almost the same as a Vector3.

**Some other minor changes**
- Fixed error when decoding CFrames, `Invalid number of arguments: 13`, caused by passing the pointer returned by `string.unpack` into `CFrame.new`. Done by assigning a variable to each component.
- Added a guard clause into `getRefId` rather than a generic if statement (probably should revert, feels less readable)
- Changed some string interpolation quotes out for double quotes (no string interpolation was being done).

**Benchmarks for encoding/decoding CFrames**
![image](https://user-images.githubusercontent.com/132710510/236658352-0a268bec-eef1-4768-8ff9-a4bebb23e99d.png)

**Test Script**
Ttests the special cases then tests random cases
```lua
local Refcoder = require(game.ReplicatedStorage.Modules.Refcoder)

local RAD_90 = math.rad(90)
local RAD_180 = math.rad(180)

local rotations = {
	CFrame.Angles(0, 0, 0),			CFrame.Angles(0, RAD_180, 0),
	CFrame.Angles(RAD_90, 0, 0),		CFrame.Angles(-RAD_90, -RAD_180, 0),
	CFrame.Angles(0, RAD_180, RAD_180),	CFrame.Angles(0, 0, RAD_180),
	CFrame.Angles(-RAD_90, 0, 0),		CFrame.Angles(RAD_90, RAD_180, 0),
	CFrame.Angles(0, RAD_180, RAD_90),	CFrame.Angles(0, 0, -RAD_90),
	CFrame.Angles(0, RAD_90, RAD_90),	CFrame.Angles(0, -RAD_90, -RAD_90),
	CFrame.Angles(0, 0, RAD_90),		CFrame.Angles(0, -RAD_180, -RAD_90),
	CFrame.Angles(0, -RAD_90, RAD_90),	CFrame.Angles(0, RAD_90, -RAD_90),
	CFrame.Angles(-RAD_90, -RAD_90, 0),	CFrame.Angles(RAD_90, RAD_90, 0),
	CFrame.Angles(0, -RAD_90, 0),		CFrame.Angles(0, RAD_90, 0),
	CFrame.Angles(RAD_90, -RAD_90, 0),	CFrame.Angles(-RAD_90, RAD_90, 0),
	CFrame.Angles(0, RAD_90, RAD_180),	CFrame.Angles(0, -RAD_90, RAD_180),
}

local checksPassed = 0
local checkCount = 100000

local encodeTime = 0
local decodeTime = 0

local function checkCFrame(cframe: CFrame)
	local startTime = os.clock()
	local encoded = Refcoder.encode(cframe)
	local middleTime = os.clock()
	local decoded = Refcoder.decode(encoded)
	local endTime = os.clock()

	if cframe == decoded then
		encodeTime += middleTime - startTime
		decodeTime += endTime - middleTime
		checksPassed += 1
	else
		warn(`Expected {cframe}, got {decoded}`)
	end
end

for _, rotation in rotations do
	local cframe = CFrame.new(math.random() * 25, math.random() * 25, math.random() * 25) * rotation
	checkCFrame(cframe)
end

for _ = 1, checkCount - #rotations do
	local cframe = CFrame.new(math.random() * 25, math.random() * 25, math.random() * 25)
		* CFrame.Angles(math.random() * 25, math.random() * 25, math.random() * 25)
	checkCFrame(cframe)
end

print(`Passed {checksPassed}/{checkCount} checks.`)
print(`Took {encodeTime / checksPassed * 1000000}μs to encode on average.`)
print(`Took {decodeTime / checksPassed * 1000000}μs to decode on average.`)
```

ArvidSilverlock committed:
Refcoder CFrame Cases

Fixed error when decoding CFrames, `Invalid number of arguments: 13`, caused by pausing the pointer returned by `string.unpack`.

Implemented CFrame special cases (https://dom.rojo.space/binary.html#cframe) to reduce string sizes, done via adding extra type ids not present in `VALID_TYPES`, each representing a CFrame's rotation.

Also added a guard clause into `getRefId ` rather than a generic if statement and changed some string interpolation quotes out for double quotes (no string interpolation was being done)

ArvidSilverlock committed:
Changed CFrame Special Case Method

Rather than changing the type id (now that I am more sleep deprived, that seems illogical and inelegant), I add an extra byte to define the case. When it's said value is 0, rotation is stored, otherwise only the position is stored.

Technically, uses like an extra character, but is far more logical and concise (no more CFrame code hanging around outside of the CFrame area).

Hexcede commented:
The lack of a guard clause when initializing `refId` is actually a style consistency choice. I use guard clauses in almost all other cases, but when initializing values from table keys I tend to prefer not using a guard clause and instead initializing the value as I do originally since a lot of the time the value gets immediately used (it's just that in this particular case, it doesn't get used).

I believe in the `metadata-refactor` branch `refId` does get used after the if statement now (although I didn't check so I might be wrong).

Hexcede commented on src/init.luau:
```diff
@@ -79,31 +98,32 @@ end
+		state.EncodingsToRefs[packed] = refId
 	end
+	
+	state.Encodings[refId] = packed
```
Should revert the changes here back to the original, since it's a conscious style choice.

Hexcede commented on src/init.luau:
```diff
@@ -54,6 +54,25 @@ local function decodeRefTable(data: string, readStart: number?)
 	return indices, values, pointer
 end
 
+local RAD_90 = math.rad(90)
+local RAD_180 = math.rad(180)
+
+-- https://dom.rojo.space/binary.html#cframe
+local CFRAME_SPECIAL_CASES = {
+	CFrame.Angles(0, 0, 0);			CFrame.Angles(0, RAD_180, 0);
+	CFrame.Angles(RAD_90, 0, 0);		CFrame.Angles(-RAD_90, -RAD_180, 0);
+	CFrame.Angles(0, RAD_180, RAD_180);	CFrame.Angles(0, 0, RAD_180);
+	CFrame.Angles(-RAD_90, 0, 0);		CFrame.Angles(RAD_90, RAD_180, 0);
+	CFrame.Angles(0, RAD_180, RAD_90);	CFrame.Angles(0, 0, -RAD_90);
+	CFrame.Angles(0, RAD_90, RAD_90);	CFrame.Angles(0, -RAD_90, -RAD_90);
+	CFrame.Angles(0, 0, RAD_90);		CFrame.Angles(0, -RAD_180, -RAD_90);
+	CFrame.Angles(0, -RAD_90, RAD_90);	CFrame.Angles(0, RAD_90, -RAD_90);
+	CFrame.Angles(-RAD_90, -RAD_90, 0);	CFrame.Angles(RAD_90, RAD_90, 0);
+	CFrame.Angles(0, -RAD_90, 0);		CFrame.Angles(0, RAD_90, 0);
+	CFrame.Angles(RAD_90, -RAD_90, 0);	CFrame.Angles(-RAD_90, RAD_90, 0);
+	CFrame.Angles(0, RAD_90, RAD_180);	CFrame.Angles(0, -RAD_90, RAD_180);
+}
+
```
Should consider generating this table automatically.

ArvidSilverlock replied:
This table is not all of the 90° snapped rotations, it's specifically the ones that `BasePart` will set their rotations to, for example, setting a parts `CFrame` to '180, 0, 0' actually sets it to `0, 180, 180`.

I'm unsure of the rhyme or reason behind these values, and thus cannot generate it automatically.

Hexcede commented on src/init.luau:
```diff
@@ -125,7 +145,9 @@ function encode(value: any, state: EncodeState, isStatic: boolean?)
-		return string.pack(`<{string.rep("nnn", 4)}`, value:GetComponents()), typeIndex
+		local caseId = table.find(CFRAME_SPECIAL_CASES, value.Rotation)
+		local packFormat = string.rep("nnn", if caseId == nil then 4 else 1)
+		return string.pack(`<B<{packFormat}`, caseId or 0, value:GetComponents()), typeIndex
```
Second `<` should be removed since it's unnecessary.

Hexcede commented on src/init.luau:
```diff
@@ -125,7 +145,9 @@ function encode(value: any, state: EncodeState, isStatic: boolean?)
 		return string.pack("<nnn", value.X, value.Y, value.Z), typeIndex
 	elseif typeof(value) == "CFrame" then
-		return string.pack(`<{string.rep("nnn", 4)}`, value:GetComponents()), typeIndex
+		local caseId = table.find(CFRAME_SPECIAL_CASES, value.Rotation)
```
Right now `caseId` doesn't really do anything beneficial, since below the CFrame's full components list is still saved anyways, so I am not entirely sure I understand what the point of it is. I am making a guess that you originally meant to conditionally save the components.

This should save only the `caseId` byte if the CFrame is in the special cases list and three numbers for the position. If it isn't (`caseId` 0) the entire CFrame's components should get saved with the byte as just `0`.

This seems to already be accounted for in the code below.

ArvidSilverlock replied:
The `packFormat` variable (defined as `string.rep("nnn", if caseId == nil then 4 else 1)`) would be "nnnnnnnnnnnn" if `caseId` is nil (meaning it is not a special CFrame case), whereas it would be "nnn" if `caseId` *isn't* nil, which means it would only store the position, and not the rotation matrix.

The reason I still store `caseId`, even if the full list of components is still present, is because I am unsure on how I would identify that it is a normal CFrame with all components listed when decoding.

Hexcede commented on src/init.luau:
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
 	elseif valueType == "CFrame" then
-		return CFrame.new(string.unpack(`<{string.rep("nnn", 4)}`, value))
+		local cframeCase, pointer = string.unpack("<B", value)
+		if cframeCase ~= 0 then
```
Should flip the branches in this if statement so the condition is positive.

Hexcede commented on src/init.luau:
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
+		local cframeCase, pointer = string.unpack("<B", value)
+		if cframeCase ~= 0 then
+			local x, y, z = string.unpack("<nnn", value, pointer)
+			return CFrame.new(x, y, z) * CFRAME_SPECIAL_CASES[cframeCase]
```
Instead of CFrame multiplication, let's use this instead:
```lua
local position = Vector3.new(string.unpack("<nnn", value, pointer))
local orientation = CFRAME_SPECIAL_CASES[cframeCase]
return CFrame.fromMatrix(position, orientation.XVector, orientation.YVector, orientation.ZVector)
```

Hexcede commented on src/init.luau:
```diff
@@ -165,13 +189,20 @@ local function decode(refId: number, encoded: string, state: DecodeState)
+			local x, y, z = string.unpack("<nnn", value, pointer)
+			return CFrame.new(x, y, z) * CFRAME_SPECIAL_CASES[cframeCase]
+		else
+			local x, y, z, R00, R01, R02, R10, R11, R12, R20, R21, R22 = string.unpack(`<{string.rep("nnn", 4)}`, value, pointer)
```
Maybe consider using `CFrame.fromMatrix` here as well, and loading `XVector`, `YVector`, and `ZVector` separately instead? I think this would be cleaner and more readable. I believe this would require serializing position, `XVector`, `YVector`, and `ZVector` above since I don't think the components are in the same order (pretty sure R00 R10 R20 makes up XVector instead of R00 R01 R02 but I might be wrong, should definitely check this)

ArvidSilverlock replied:
I'm not too sure how I would achieve this whilst retaining the same performance, as if i'm not wrong, you want construct 4 `Vector3`s (whilst unpacking 4 separate times?)


According to the [CFrame docs](https://create.roblox.com/docs/reference/engine/datatypes/CFrame), R00, R10 and R20 make up to RightVector, but the XVector is made up of R00, R01 and R02.
![image](https://user-images.githubusercontent.com/132710510/236703626-2c9f9233-9739-4ba8-a73d-a833cafe48f7.png)

ArvidSilverlock committed:
Implemented Some Requested Changes

- Reverted refid guard clause back to a generic if-statement.
- Removed unnecessary `<`.
- Flipped the decode branches for CFrames.
- Swapped to `CFrame.fromMatrix` when decoding special CFrame cases.
- Changed `string.rep` out for hardcoded strings, in an attempt to micro-optimise.

Benchmarked at ~6.12μs to encode, 2.72μs to decode over 10,000,000 random CFrames.

Ignore the indentation changes, that's roblox's formatter, any further changes I make won't have them.
